### DISCUSSION TITLE
Multi-Model pattern clarification (UnionDoc)

### DISCUSSION THREAD

**Discussion by Vitalium** [BeanieODM/beanie#366] [Question] [answered]
Hello!

The [docs about UnionDoc](https://beanie-odm.dev/tutorial/multi-model-pattern/) is quite superficial. There are many questions and I hope to get answers that will help me to understand the subject more deeply as well as to improve documentation.

1) `class Parent(UnionDoc)` in example has no attributes defined, while in classes `One` & `Two` there is `shared: int = 0` which is shared between two sub-instances. It is more obvious to take out this field to the `Parent`. I see that the latter has `UnionDoc` type, that is not derived from BaseModel or Document. So there is a question: can `Parent: UnionDoc` store attributes (fields)?

2) What if I need to use `Parent` as standalone objects on a par with `One` and `Two`? In the docs there is an option to make query on `Parent` to get both "subclasses". In my case there are following entities:
```python
class User(UnionDoc):
    ...


class Worker(Document):
    class Settings:
        union_doc = User


class Admin(Document):
    class Settings:
        union_doc = User
```

`User` is not just an abstraction, it is full-fledged entity. `Worker` and `Admin` should be derived from `User` and have the same fields and also extended with some specific fields. I guess it this case declarations should be like this:

```python
class User(UnionDoc):
    ...


class Customer(Document):
    name: str
    settings: dict[str, Any]

    class Settings:
        union_doc = User


class Worker(Customer):
    salary: float
    departments: list[Link['Dept']]

    class Settings:
        union_doc = User


class Admin(Worker):
    some_field: str
    subordinates: list[Link[Worker]] = []
    class Settings:
        union_doc = User
```
Using above example querying User will return a collection on all types of children. But what if I query on `Customer`? Will I get `Workers` and `Admins` too (cause they are derived from the `Customer`)? And how can I filter a query to get only `Customers` without objects of children subclasses?

2.1) I also have another type of classes that, on the contrary, require an abstract entity approach. But the same as above I need to store shared fields somewhere. I guess I can use shared `pydantic.BaseModel`.

```python
class Server(UnionDoc):
    ...


class ServerModel(BaseModel):
    name: str
    price: float
    ip: IPv4Address | None = None
    ordered: datetime | None = None
    user: Link[User] | None = None


class PhysicalServer(ServerModel, Document):
    # physical location, rack name and unit number
    location: str
    ipmi: Link['IPMI'] | None = None
    # connection data to the external devices like Switch, PDU (Power Distribution Unit), IP-KVM etc.
    connections: list[Link['Connection']] = []

    class Settings:
        union_doc = Server


class VirtualServer(ServerModel, Document):
    # virtualization type
    hypervisor: str
    uuid: UUID

    class Settings:
        union_doc = Server    
```
In this case an abstract `Server` cannot be created, only one of two type, so `UnionDoc` seems like a perfect fit. If the above example is correct, then this is the answer to question #1.

3) Can I just use standard pythonic inheritance instead of UnionDoc approach? Will child classes be stored in the same collection? If not, what if I specify the same collection name in the `Settings` inner class?

**Comment by roman-right** [comment, marked as answer]
Hey!
Sorry for the late reply. 

1. `UnionDoc` is a kind of controller (not model) in this scope. It has no data fields in this case.
2. The example from 2.1 should work and it reflects the idea I had implementing this pattern.
3. Sure. It will be stored in the same collection if no specific collection name is added to the Settings class. This `UnionDoc` class was added for cases, where there are conflict fields in the models but things like aggregations should work against a single collection. But people use it for other scenarios too :)

**Reply by mg3146** [reply]
In the case of 2.1, can I instead of BaseModel use a Document to inherit from?  Or can a BaseModel include things like "Indexed()" parameters?

**Comment by mg3146** [comment]
I would love an expanded section on the UnionDoc as well.  I just posted on it, but I'm getting a UnionDocNotInited error, and can't really figure out why.

**Reply by mg3146** [reply]
For example, if I use your 2.1 as a guide.  Can I instead replace the BaseModel with a Document? I'm trying to have all my subclasses inherit event-based actions, etc.

**Reply by Vitalium** [reply]
I think you can, but that `Document` probably should not to be registered upon init.

**Reply by mg3146** [reply]
Got you - the init here you're talking about is the init_beanie() ? tx!

**Comment by Vitalium** [comment]
Yes init_beanie, but I'm not completely sure about it, try to test youself.

**Comment by gsakkis** [comment]
I'm also confused about `UnionDoc` vs inheritance:
- @roman-right [mentions](https://github.com/roman-right/beanie/discussions/366#discussioncomment-3795646) `UnionDoc is a kind of controller (not model) in this scope`. Not exactly sure what this means but it contradicts the docs that refer to "controller" in the inheritance section:
    > This behavior is similar to UnionDoc, but you don't need an additional entity. Parent Document act like a "controller", that handles proper storing and fetches different Document types.
- My understanding based on the docs is that inheritance provides a superset of functionality of `UnionDoc`; anything that `UnionDoc` does can be done by inheritance but not the other way around (e.g. `UnionDoc` has no data fields to be shared). So either:
  - my understanding is wrong, in which case the docs should expand on the differences and help the user decide how to pick between the two approaches
  - or `UnionDoc` is technically redundant, in which case it could be candidate for deprecation/removal for v2.

**Reply by roman-right** [reply]
Hi! With UnionDoc you tie the models (they could be completely different) with a controller that can be used to query these models. Inheritance defines another relationship - it is a kind of hierarchy. I think both could be useful for different use cases. I don't plan to deprecate any of them for now. If I'll decide, I'll add a deprecation warning.

**Reply by gsakkis** [reply]
Thanks! An updated example in the docs that demonstrate a use case where `UnionDoc` is either the only or preferred option would be great. The current example even uses a class name that suggests inheritance (`union_doc = Parent`).
